lib/FusionInventory/Agent/Task/Inventory/Linux/Storages/Megacli2.pm

```diff
@@ -0,0 +1,145 @@
+package FusionInventory::Agent::Task::Inventory::Linux::Storages::Megacli2;
+
+use strict;
+use warnings;
+
+use parent 'FusionInventory::Agent::Task::Inventory::Module';
+
+use File::Basename qw(basename);
+use FusionInventory::Agent::Tools;
+use FusionInventory::Agent::Tools::Linux qw(getInfoFromSmartctl);
+
+use constant RE => qr/^([^:]+?)\s*:\s*(.*\S)/;
+
+sub isEnabled {
+    return canRun('megacli') && canRun('smartctl');
+}
+
+sub doInventory {
+    my (%params) = @_;
+
+    my $inventory = $params{inventory};
+
+    my $count = getFirstMatch(
+        command => "megacli -adpCount",
+        pattern => qr/Controller Count: (\d+)/
+    );
+    return unless $count;
+
+    for (my $adp = 0; $adp < $count; $adp++) {
+        my $adpinfo = _getAdpPciInfo(adp => $adp);
+        my $block = _adpInfoToBlock($adpinfo);
+        my $pdlist = _getPDlist(adp => $adp);
+        my $ldinfo = _getLDinfo(adp => $adp);
+
+        while (my ($id, $pd) = each %$pdlist) {
+            # JBOD and Non-RAID are processed by the parent module, skip.
+            next if (
+                (defined $pd->{__diskgroup} &&
+                 defined $ldinfo->{ $pd->{__diskgroup} } &&
+                 defined $ldinfo->{ $pd->{__diskgroup} }->{'Number Of Drives'} &&
+                 defined $ldinfo->{ $pd->{__diskgroup} }->{'Name'} &&
+                 $ldinfo->{ $pd->{__diskgroup} }->{'Number Of Drives'} eq '1' &&
+                 $ldinfo->{ $pd->{__diskgroup} }->{'Name'} =~ /Non\s*\-*RAID/i)
+                    ||
+                $pd->{'Firmware state'} =~ /JBOD/i
+            );
+
+            my $storage = getInfoFromSmartctl(
+                device => '/dev/' . $block,
+                extra  => '-d megaraid,' . $id);
+
+            $inventory->addEntry(
+                section => 'STORAGES',
+                entry   => $storage,
+            );
+        }
+    }
+}
+
+sub _getPDlist {
+    my (%params) = @_;
+
+    $params{command} = defined $params{adp} ? "megacli -pdlist -a$params{adp}" : undef;
+
+    my (%srclist, %pdlist);
+    my $n = 0;
+
+    foreach my $line (getAllLines(%params)) {
+        next unless $line =~ RE;
+        my ($key, $val) = ($1, $2);
+        $n++ if $key =~ /Enclosure Device ID/;
+        $srclist{$n}->{$key} = $val;
+    }
```

Comment:
```suggestion
    my (@srclist, %pdlist);
    my $src = {};

    foreach my $line (getAllLines(%params)) {
        my ($key, $val) = $line =~ RE
            or next;
        if ($key =~ /Enclosure Device ID/) {
            push @srclist, $src;
            $src = {};
        } else {
            $src->{$key} = $val;
        }
    }
    push @srclist, $src;
```
Can you better use a list as this is more efficient than managing a %srclist hash ? It's always a pain to re-read a code when using a hash where an array is sufficient as we always try to understand why a hash is used. Another point when using values on a hash is we don't get values in the same order. I don't think this would be a problem for this case, but this can lead us to other problems if the final primary key (here 'Device Id') can be the same on few list elements.
If you adopt my suggestion, don't forget to update l.75 accordingly. Btw I think you can even merge the 2 loops by directly updating the `%pdlist` and remove the need of `@srclist`:
* replace all the `push` by `$pdlist{$src->{'Device Id'}} = $src if defined($src->{'Device Id'});`
* add `$src->{__diskgroup}`, `$src->{__span}`, `$src->{__arm}` updates into the `else` part `if $key eq  "Drive's position"` and applying the regexp on `$val`

Reply:
The primary reason for using hashes instead of arrays is that one doesn't need that extra duplicate code to process whatever is left in a temporary hash after `for` loop ends. `push @srclist, $src;` in your case.

Here's the code you suggested, but i absolutely dislike the last line :D . Suggestions?

```perl
    foreach my $line (getAllLines(%params)) {
        my ($key, $val) = $line =~ RE
            or next;

        if ($key eq 'Enclosure Device ID') {
            $pdlist{ $src->{'Device Id'} } = $src if defined $src->{'Device Id'};
            $src = {};
        } elsif ($key eq 'Drive\'s position') {
            ($src->{__diskgroup}, $src->{__span}, $src->{__arm}) =
                ($val =~ /DiskGroup: (\d+), Span: (\d+), Arm: (\d+)/);
        }

        $src->{$key} = $val;
    }
    $pdlist{ $src->{'Device Id'} } = $src if defined $src->{'Device Id'};
```

Reply:
I swallowed it and committed. :)